File: Problem1.java
//# Design-4

//## Problem 1: Design Twitter (https://leetcode.com/problems/design-twitter/)

class Twitter {
    class Tweet{
        public int id;
        public int timestamp;

        public Tweet(int id, int time){
            this.id=id;
            this.timestamp=time;
        } 
    }

    // Map to store the ids of users, current user is following
    private Map<Integer, Set<Integer>> following;
    // Map to store the tweets of a certain user
    private Map<Integer, List<Tweet>> tweets;
    // Keep track of global time
    private int time;

    public Twitter() {
        this.following=new HashMap<>();
        this.tweets=new HashMap<>();
        this.time=0;
    }
    
    // Time O(1)
    public void postTweet(int userId, int tweetId) {
        // Follow self to get self-tweets
        follow(userId, userId);
        if(!tweets.containsKey(userId)){
            // First tweet of a user
            tweets.put(userId, new ArrayList<>());
        }
        tweets.get(userId).add(new Tweet(tweetId, this.time));
        this.time=this.time+1;
    }
    
    // Time: O(n log(k)), where n is the total number of following of a user and k=10
    // Space: O(k) for the priority queue
    public List<Integer> getNewsFeed(int userId) {
        List<Integer> result=new ArrayList<>();

        //1. Get all the followings of current user --- O(1)
        if(!following.containsKey(userId)){
            // User doesn't follow anyone and has no tweets of their own
            return result;
        }
        Set<Integer> userFollowing = following.get(userId);

        //2. Get the most recent k tweets from the users, current user is following
        // Put them one by one in a min-heap of size 10
        // Time - O(nk logk), where n is the total number of following of a user and k=10
        PriorityQueue<Tweet> pq=new PriorityQueue<>((a,b) -> a.timestamp - b.timestamp);
        for(Integer fId: userFollowing){
            List<Tweet> fTweets = tweets.getOrDefault(fId, new ArrayList<>());
            
            for(int i=fTweets.size()-1;i>=0 && i>=(fTweets.size()-10);i--){
                pq.add(fTweets.get(i));
                if(pq.size()>10){
                    pq.poll();
                }
            }

        }

        //3. Pop elements from heap one by one and add to result
        // Since we are using min-heap, the top will not represent oldest of the 10 tweets
        // We need to store the result in reverse
        // Time - O(k), where k is 10
        while(!pq.isEmpty()){
            result.add(0, pq.poll().id);
        }

        return result;
    }
    
    // Time O(1)
    // Space O(1)
    public void follow(int followerId, int followeeId) {
        if(!following.containsKey(followerId)){
            // First time following someone
            following.put(followerId, new HashSet<>());
        }
        following.get(followerId).add(followeeId);
    }
    
    // Time O(1)
    // Space O(1)
    public void unfollow(int followerId, int followeeId) {
        if(!following.containsKey(followerId)){
            // User doesn't follow anyone
            return;
        }
        following.get(followerId).remove(followeeId);
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: Problem_1.py
'''
Time Complexity - O(1) for follow, unfollow and tweet. O(nlogk) for getting feed
Space Complexity - O(n) for storing the userId-> followees, userId->tweets, O(log k) for getting tweets  

Works on LeetCode.
'''


import heapq
class Tweet:
    def __init__(self, tweetId, createdAt):
        self.tweetId = tweetId #store the TweetId
        self.createdAt = createdAt #store the timestamp of the tweet

    def __lt__(self, tweet):
        return self.createdAt < tweet.createdAt

class Twitter:

    def __init__(self):
        self.followedMap = {} #Follower(UserId)->[Followees]
        self.tweetsMap = {} #UserID->[TweetIds]
        self.createdAt = 0 #maintain Time

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.follow(userId, userId) #make a user follow itself so that it can its own tweets
        if userId not in self.tweetsMap: #if user has not tweeted before
            self.tweetsMap[userId] = [] #create an entry
        tweets = self.tweetsMap[userId] #get the list of tweets for the user
        tweets.append(Tweet(tweetId, self.createdAt)) #create a new tweet, enter time of creation, add to list of tweets of userId
        self.createdAt+=1


    def getNewsFeed(self, userId: int) -> List[int]:
        pq = [] #initiate the heap
        if userId not in self.followedMap:
            return [] #return empty list if the user does not follow anyone
        followed = self.followedMap[userId] #get the list of users the user follows
        result = []
        count = 0
        if len(followed)>0:
            for fid in followed:
                

                ftweets = self.tweetsMap.get(fid) if fid in self.tweetsMap else [] #get the tweets for every user the user follows
                if len(ftweets)>0:
                    for ftweet in ftweets: 
                        heapq.heappush(pq,ftweet)#push each tweet in the heap
                        count+=1 #increase the size of heap
                        if count > 10: #we only need the 10 latest tweets so we pop all the tweets with earlier times
                            heapq.heappop(pq) 
                            count-=1# while popping we also reduce the size by 1
        while count > 0 :
            tw = heapq.heappop(pq)
            result.append(tw.tweetId) #in ther end we append the latest 10 tweets to a list
            count-=1
        return result[::-1]       #we reverse and return the list           
                  
        

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.followedMap: #if the user does not follow anyone or is new, 
            self.followedMap[followerId] = set() #the user is created, and a set of followees for the user is created
        followersSet = self.followedMap[followerId] 
        followersSet.add(followeeId) #add the followee to the set of followees

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.followedMap[followerId] and followerId != followeeId: #if the followee is present in the list of followees and the followee id is not the same as that of the user
            followersSet = self.followedMap[followerId] #get the list of followees
            followersSet.remove(followeeId) #remove the followee from the list



# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)